kong/runloop/wasm.lua

```diff
@@ -0,0 +1,646 @@
+local function build_filter_list(service_chain, route_chain)
+  ---@type kong.db.schema.entities.wasm_filter[]|nil
+  local combined
+  local n = 0
+
+  if service_chain and service_chain.enabled then
+    for _, filter in ipairs(service_chain.filters) do
+      if filter.enabled then
+        n = n + 1
+        combined = combined or {}
+        combined[n] = filter
+      end
+    end
+  end
+
+  if route_chain and route_chain.enabled then
+    for _, filter in ipairs(route_chain.filters) do
+      if filter.enabled then
+        n = n + 1
+        combined = combined or {}
+        combined[n] = filter
+      end
+    end
+  end
+
+  return combined
+end
+
+
+---
+-- Unconditionally rebuild and return a new wasm state table from the db.
+--
+---@param  db                       table # kong.db
+---@param  version                  any
+---@param  old_state                kong.runloop.wasm.state
+---@return kong.runloop.wasm.state? new_state
+---@return string?                  err
+local function rebuild_state(db, version, old_state)
+  ---@type kong.db.schema.entities.filter_chain[]
+  local route_chains = {}
+
+  ---@type table<string, kong.db.schema.entities.filter_chain>
+  local service_chains_by_id = {}
+
+  ---@type kong.runloop.wasm.state
+  local state = {
+    by_service = {},
+    by_route = {},
+    combined = {},
+    version = version,
+  }
+
+  ---@type kong.runloop.wasm.filter_chain_reference[]
+  local all_chain_refs = {}
+
+  local page_size = db.filter_chains.max_page_size
+
+  for chain, err in db.filter_chains:each(page_size) do
+    if err then
+      return nil, "failed iterating filter chains: " .. tostring(err)
+    end
+
+    if chain.enabled then
+      local route_id = chain.route and chain.route.id
+      local service_id = chain.service and chain.service.id
+
+      local chain_type = service_id and TYPE_SERVICE or TYPE_ROUTE
+
+      insert(all_chain_refs, {
+        type           = chain_type,
+
+        service_chain  = (chain_type == TYPE_SERVICE and chain) or nil,
+        service_id     = service_id,
+
+        route_chain    = (chain_type == TYPE_ROUTE and chain) or nil,
+        route_id       = route_id,
+      })
+
+      if chain_type == TYPE_SERVICE then
+        service_chains_by_id[service_id] = chain
+
+      else
+        insert(route_chains, chain)
+      end
+    end
+  end
+
+  local routes = db.routes
+  local select_route = routes.select
+
+  -- the only cache lookups here are for route entities,
+  -- so use the core cache
+  local cache = kong.core_cache
+
+
+  -- locate matching route/service chain entities to build combined
+  -- filter chain references
+  for _, rchain in ipairs(route_chains) do
+    local cache_key = routes:cache_key(rchain.route.id)
+
+    local route, err = cache:get(cache_key, nil,
+                                 select_route, routes, rchain.route)
+
+    if err then
+      return nil, "failed to load route for filter chain " ..
```

Comment:
ditto, not sure if "wasm" will be concatenated to this error string by a caller or not, but this string does not mention "wasm" at all.